Allow TCPStore to pick a port to bind to.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31674 Allow TCPStore to pick a port to bind to.**

The motivation of this PR was to fix the problem where we would see
"Address already in use" issues for TCPStoreTest due to port conflicts. To
resolve this:

1. We can now pass in port 0 for TCPStore and retrieve the port it actually
bound to using a new getPort() API.
2. Added a `wait` flag to TCPStore constructor indicating whether or not it
should wait for workers (defaults to true).
3. Made `waitForWorkers` a public API to ensure that we can construct TCPStore
without waiting and wait for workers separately. This helps in TCPStoreTest to
ensure we can retrieve the port and pass it to the client stores.

Differential Revision: [D19240947](https://our.internmc.facebook.com/intern/diff/D19240947/)

pritamdamania committed:
Allow TCPStore to pick a port to bind to.

The motivation of this PR was to fix the problem where we would see
"Address already in use" issues for TCPStoreTest due to port conflicts. To
resolve this:

1. We can now pass in port 0 for TCPStore and retrieve the port it actually
bound to using a new getPort() API.
2. Added a `wait` flag to TCPStore constructor indicating whether or not it
should wait for workers (defaults to true).
3. Made `waitForWorkers` a public API to ensure that we can construct TCPStore
without waiting and wait for workers separately. This helps in TCPStoreTest to
ensure we can retrieve the port and pass it to the client stores.

Differential Revision: [D19240947](https://our.internmc.facebook.com/intern/diff/D19240947/)

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 2b184463:


**None of the build failures appear to be your fault**.

* **1/1** recognized as flaky :snowflake:
    * Re-run these jobs?


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=2b1844630c128b1e19e3782d6bf3874d42035b9e).

### :snowflake: 1 failure recognized as flaky
The following build failures have been detected as flaky and **may not be your fault**:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4078799) pytorch_linux_xenial_cuda9_cudnn7_py3_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=78734469) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=78734469)) :snowflake:
<details>
<summary>
<code>Dec 28 02:06:05 AssertionError: 1024 not less than or equal to 1e-05 : __main__.TestAutogradDeviceTypeCUDA.test_logdet_1x1_cuda leaked 1024 bytes CUDA memory on device 0</code>
</summary>

```
Dec 28 02:06:05 ====================================================================== 
Dec 28 02:06:05 FAIL [0.102s]: test_logdet_1x1_cuda (__main__.TestAutogradDeviceTypeCUDA) 
Dec 28 02:06:05 ---------------------------------------------------------------------- 
Dec 28 02:06:05 Traceback (most recent call last): 
Dec 28 02:06:05   File "/var/lib/jenkins/workspace/test/common_utils.py", line 665, in wrapper 
Dec 28 02:06:05     method(*args, **kwargs) 
Dec 28 02:06:05   File "/var/lib/jenkins/workspace/test/common_utils.py", line 521, in __exit__ 
Dec 28 02:06:05     self.name, after - before, i)) 
Dec 28 02:06:05   File "/var/lib/jenkins/workspace/test/common_utils.py", line 877, in assertEqual 
Dec 28 02:06:05     super(TestCase, self).assertLessEqual(abs(x - y), prec, message) 
Dec 28 02:06:05 AssertionError: 1024 not less than or equal to 1e-05 : __main__.TestAutogradDeviceTypeCUDA.test_logdet_1x1_cuda leaked 1024 bytes CUDA memory on device 0 
Dec 28 02:06:05  
Dec 28 02:06:05 ---------------------------------------------------------------------- 
Dec 28 02:06:05 Ran 1884 tests in 880.219s 
Dec 28 02:06:05  
Dec 28 02:06:05 FAILED (failures=1, skipped=13, expected failures=1) 
Dec 28 02:06:05  
Dec 28 02:06:05 Generating XML reports... 
Dec 28 02:06:05 Traceback (most recent call last): 
Dec 28 02:06:05   File "test/run_test.py", line 456, in <module> 
Dec 28 02:06:05     main() 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

satgera commented on torch/lib/c10d/TCPStore.cpp:
```diff
@@ -273,7 +273,8 @@ TCPStore::TCPStore(
     bool isServer,
-    const std::chrono::milliseconds& timeout)
+    const std::chrono::milliseconds& timeout,
+    bool wait)
```
Nit: may be call expand the name to something more meaningful so that the caller knows what will it wait for with the flag ?

pritamdamania87 replied:
Makes sense, will address before landing.

satgera reviewed (approved):
Thanks for doing this, looks good. Ensure no failure is related.

pritamdamania87 commented:
@pietern Could you take a look at this PR too? There are a few API changes to TCPStore here.

pietern reviewed (approved):
LGTM.

I think the optional bool to wait or not wait could be more clear with an enum, but it's a minor thing.

pritamdamania committed:
Update on "Allow TCPStore to pick a port to bind to."


The motivation of this PR was to fix the problem where we would see
"Address already in use" issues for TCPStoreTest due to port conflicts. To
resolve this:

1. We can now pass in port 0 for TCPStore and retrieve the port it actually
bound to using a new getPort() API.
2. Added a `wait` flag to TCPStore constructor indicating whether or not it
should wait for workers (defaults to true).
3. Made `waitForWorkers` a public API to ensure that we can construct TCPStore
without waiting and wait for workers separately. This helps in TCPStoreTest to
ensure we can retrieve the port and pass it to the client stores.

Differential Revision: [D19240947](https://our.internmc.facebook.com/intern/diff/D19240947/)

[ghstack-poisoned]
